cylc/flow/cycling/iso8601.py

```diff
@@ -391,7 +391,7 @@ def set_offset(self, i_offset):
         if self.exclusions:
             self.value += '!' + str(self.exclusions)
 
-    @lru_cache(100)
+    @lru_cache(100)  # noqa B019 TODO?
```

Comment:
Need to think about this.

Note recent official Python docs advice:

[How do I cache method calls](https://docs.python.org/3/faq/programming.html#how-do-i-cache-method-calls)

Reply:
I'm happy to approve this for now, but I think that we should create an issue to investigate the context and decide which caching approach is appropriate.

AFAI can tell this advice has been added to the FAQ at 3.10, and the `functools.cached_property` was [added at 3.8](https://docs.python.org/3.8/whatsnew/3.8.html#functools) so it's not really available to us unless we manually back-port the [class](https://github.com/python/cpython/blob/7ad52d174a12951d260aecfcdfe0fcfd2437a66b/Lib/functools.py#L966) in our code base. 🤮

Reply:
I don't think this usage is an issue for us as on the whole Cylc doesn't make a habit of releasing `ISO8601Sequence` instances (they are used by taskdefs which live for the life of the config).

The `cached_property` thing can't handle arguments so doesn't help, the issue is that the cache lives for the life of the class/global scope rather than the life of the instance, it looks like there is an easy way to solve that.

Reply:
> The lru_cache approach can be made to work, but the class needs to define the __eq__ and __hash__ methods so the cache can detect relevant attribute updates:

https://github.com/cylc/cylc-flow/blob/85694062b09d10e64541f76835c5507f1c8a0895/cylc/flow/cycling/iso8601.py#L589-L604

Reply:
(Off topic but it looks to me like the `__hash__` method should `return hash((self.TYPE, self.value))` based on the `__eq__`

Reply:
Got a fix on https://github.com/cylc/cylc-flow/pull/4747

Reply:
> Off topic but it looks to me

I don't think it need to because `self.TYPE` is a class variable and the cache "belongs" to the class so `self.TYPE` should not differ.

The reason that the `__eq__` method does this is to allow comparisons with other sequence types.